sim/core/resource_bar.go

```diff
@@ -0,0 +1,183 @@
+	bar.value = 0
+	if bar.config.Default > 0 {
+		bar.Gain(bar.config.Default, ActionID{SpellID: int32(bar.config.Type)}, sim)
+	}
+}
+
+// Spend implements SecondaryResourceBar.
+func (bar *DefaultSecondaryResourceBarImpl) Spend(amount int32, action ActionID, sim *Simulation) {
+	if amount > bar.value {
+		panic("Trying to spend more resource than is available.")
+	}
+
+	if amount < 0 {
+		panic("Trying to spend negative amount.")
+	}
+
+	metrics := bar.GetMetric(action)
+	if sim.Log != nil {
+		bar.unit.Log(
+			sim,
+			"Spent %0.0f generic resource from %s (%0.0f --> %0.0f) of %0.0f total.",
+			amount,
+			metrics.ActionID,
+			bar.value,
+			bar.value-amount,
+			bar.config.Max,
+		)
+	}
+
+	metrics.AddEvent(float64(-amount), float64(-amount))
+	bar.invokeOnSpend(amount)
+	bar.value -= amount
+}
+
+// SpendUpTo implements SecondaryResourceBar.
+func (bar *DefaultSecondaryResourceBarImpl) SpendUpTo(limit int32, action ActionID, sim *Simulation) int32 {
+	if bar.value > limit {
+		bar.Spend(limit, action, sim)
+		return limit
+	}
+
+	bar.Spend(bar.value, action, sim)
+	return bar.value
+}
+
+// Value implements SecondaryResourceBar.
+func (bar *DefaultSecondaryResourceBarImpl) Value() int32 {
+	return bar.value
+}
+
+func (bar *DefaultSecondaryResourceBarImpl) GetMetric(action ActionID) *ResourceMetrics {
+	metric, ok := bar.metrics[action]
+	if !ok {
+		metric = bar.unit.NewGenericMetric(action)
+		bar.metrics[action] = metric
+	}
+
+	return metric
+}
+
+func (bar *DefaultSecondaryResourceBarImpl) RegisterOnGain(callback OnGainCallback) {
+	if callback == nil {
+		panic("Can not register nil callback")
+	}
+
+	bar.onGain = append(bar.onGain, callback)
+}
+
+func (bar *DefaultSecondaryResourceBarImpl) RegisterOnSpend(callback OnSpendCallback) {
+	if callback == nil {
+		panic("Can not register nil callback")
+	}
+
+	bar.onSpend = append(bar.onSpend, callback)
+}
+
+func (bar *DefaultSecondaryResourceBarImpl) invokeOnGain(gain int32, realGain int32) {
+	for _, callback := range bar.onGain {
+		callback(gain, realGain)
+	}
+}
+
+func (bar *DefaultSecondaryResourceBarImpl) invokeOnSpend(amount int32) {
+	for _, callback := range bar.onSpend {
+		callback(amount)
+	}
+}
+
+func (unit *Unit) RegisterSecondaryResourceBar(config SecondaryResourceConfig) SecondaryResourceBar {
+	if config.Type <= 0 {
+		panic("Invalid SecondaryResourceType given.")
+	}
+
+	if config.Max <= 0 {
+		panic("Invalid maximum resource value given.")
+	}
+
+	if config.Default < 0 || config.Default > config.Max {
+		panic("Invalid default value given for resource bar")
+	}
+
+	if unit.SecondaryResourceBar != nil {
+		panic("A secondary resource bar has already been registered.")
+	}
+
+	unit.SecondaryResourceBar = &DefaultSecondaryResourceBarImpl{
+		config:  config,
+		unit:    unit,
+		metrics: make(map[ActionID]*ResourceMetrics),
+		onGain:  []OnGainCallback{},
+		onSpend: []OnSpendCallback{},
+	}
+
+	return unit.SecondaryResourceBar
+}
```

Comment:
Now that I see the Paladin implementation, I realize that we do need to break this out into separate public access methods:
```
func (unit *Unit) NewDefaultSecondaryResourceBar(config SecondaryResourceConfig) *DefaultSecondaryResourceBarImpl {
```
and
```
func (unit *Unit) RegisterSecondaryResourceBar(impl SecondaryResourceBar) SecondaryResourceBar {
```
The two could be chained together with a third helper for specs that don't need any custom implementations:
```
func (unit *Unit) RegisterDefaultSecondaryResourceBar(config SecondaryResourceConfig) SecondaryResourceBar {
```